File: module_qkpack/qkpack/lib/qkpack_metrics/src/config/qkpack_acl.cc
#include "config/qkpack_acl.h"
#include "qkpack_log.h"
#include "util/string_util.h"
#include "config/easycurl.h"

using namespace com::youku::data::qkpack::config;
using namespace com::youku::data::qkpack::util;

QkPackACL::QkPackACL(std::string acl_url,int timeout):
	acl_url_(acl_url),
	timeout_(timeout)
{
	qkpack_yajl_json_ = new QkPackYajlJSON();
}


QkPackACL::~QkPackACL()
{
	if ( qkpack_yajl_json_ ) {
		delete qkpack_yajl_json_;
		qkpack_yajl_json_ = NULL;
	}

	
	std::map<std::string, qkpack_acl_t*>::const_iterator		it;

	for ( it = map_acl_.begin(); it != map_acl_.end(); ++it) {
		
		qkpack_acl_t * acl =  it->second;	
		if ( acl )  {
			delete acl;
			acl = NULL;
		}
	}
}


int QkPackACL::SetNameSpace(qkpack_request_t *request) 
{
	int                                             j,num = 0;
	const char*                                           data = (char*)request->uri.c_str();
	int                                             len = request->uri.length();
	
	if ( strncmp(data, (const char*)QKPACK_ACL_URI_FORMAT, strlen(QKPACK_ACL_URI_FORMAT)) ) {
		request->desc = QKPACK_ERROR_ACL_FORMAT;
		return QKPACK_ERROR;
	}

	for (j = 0;j < len;j++)
	{
		if ( data[j] == 47 ) num++;
		if ( num == 3 ) break;
	}

	if ( j <= 0 || num != 3 ) {
		request->desc = QKPACK_ERROR_ACL_NAMESPACE_NOT_EMPTY;
		return QKPACK_ERROR;
	}

	request->kvpair.namespaces = std::string(data+j+1);
	if ( request->kvpair.namespaces.length() != QKPACK_ACL_NAMESPACE_LENGTH ) {
		request->desc = QKPACK_ERROR_ACL_NAMESPACE_LENGTH;
		return QKPACK_ERROR;
	}
	
	return QKPACK_OK;
}

int QkPackACL::SetClusterName(qkpack_request_t *request) 
{
	int                                             j = 0,first = 0,last = 0,num = 0;
	const char*                                           data = (char*)request->uri.c_str();
	int                                             len = request->uri.length();
	
	for (j = 0;j < len;j++)
	{
		if ( data[j] == 47 ) num++;
		if ( num == 2 && first == 0 ) first = j;
		if ( num == 3 && last == 0 ) {last = j;break;}
	}
	
	if ( j <=0 || first == 0 || last == 0 ) {
		return QKPACK_ERROR;
	}

	request->cluster_name = std::string(data+first+1, last-first-1);
	return QKPACK_OK;
}	

int QkPackACL::GetZsetLimit(qkpack_request_t *request, const std::vector<limit_policies_t> &vector_policies)
{
	int								key_len;
	bool							status = false;

	if ( request->kvpair.command_type != REDIS_ZADD &&
			request->kvpair.command_type != REDIS_ZBATCHADD ) {
		return QKPACK_OK;
	}

	std::map<std::string, std::vector<kvpair_t> >				&map_kv = request->map_kv;
	std::map<std::string, std::vector<kvpair_t> >::iterator		it;

	for ( it = map_kv.begin(); it != map_kv.end(); ++it ) {

		key_len = it->first.length();

		for ( size_t i = 0; i < vector_policies.size(); ++i ) {
		
			if ( key_len >= vector_policies[i].min && key_len <= vector_policies[i].max ) {
				
				status = true;
				it->second[0].limit = vector_policies[i].limit;
				break;
			}

		}
		
		if ( !status ) {
			request->desc.append(QKPACK_ERROR_ACL_LIMITPOLICIES_LIMIT);
			request->desc.append(StringUtil::ToString(key_len));

			return QKPACK_ERROR;
		}
	}

	return QKPACK_OK;
}

//判断接口的超时时间
int QkPackACL::CheckInterfaceTimeout(qkpack_request_t *request,const std::vector<timeouts_t> &vector_timeouts)
{
	std::size_t						found;

	for ( size_t i = 0; i < vector_timeouts.size(); ++i ) {
	
		found = request->request_uri.find(vector_timeouts[i].interface_name);
		if ( found != std::string::npos ) {
		
			request->timeout = vector_timeouts[i].timeout;
			request->tries = vector_timeouts[i].tries;
			break;
		}
	}

	return QKPACK_OK;
}


//判断接口的长度限制(keylen,vallen,datalen)
int QkPackACL::CheckInterfaceLimitsLength(qkpack_request_t *request, const std::vector<limits_length_t> &vector_limits_len)
{
	std::size_t						found;
	int								keylen = request->kvpair.key.length();
	int								vallen = request->kvpair.value.length();
	int								datalen = request->request_buffer.length();

	for ( size_t i = 0; i < vector_limits_len.size(); ++i ) {
	
		found = request->request_uri.find(vector_limits_len[i].interface_name);
		if ( found != std::string::npos ) {
		
			if ( keylen &&  keylen > vector_limits_len[i].keylen ) {
				
				request->desc = QKPACK_ERROR_ACL_LIMITSLENGTH_KEYLEN;
				return QKPACK_ERROR;
			}

			if ( vallen && vallen > vector_limits_len[i].vallen ) {
				
				request->desc = QKPACK_ERROR_ACL_LIMITSLENGTH_VALLEN;
				return QKPACK_ERROR;
			}

			if ( datalen && datalen > vector_limits_len[i].datalen ) {
				
				request->desc = QKPACK_ERROR_ACL_LIMITSLENGTH_DATALEN;;
				return QKPACK_ERROR;
			}
		}
	}

	return QKPACK_OK;
}


int QkPackACL::GetNodeList(qkpack_request_t *request, qkpack_acl_t *acl)
{
	std::string												key;
	std::string												&x_real_ip = request->x_real_ip;
	std::map<std::string,std::string>						&map_server = acl->route.map_server;
	std::map<std::string,std::string>						&map_client = acl->route.map_client;
	std::map<std::string,std::string>::const_iterator		it_server;
	std::map<std::string,std::string>::const_iterator		it_client;
	char*													data = (char*)x_real_ip.c_str();
	int														len = x_real_ip.length();
	int														j,num = 0;
	char													rest[10];
	
	request->kvpair.exptime = acl->quotation.expire;
	request->kvpair.uri_id = acl->uri_id;
	request->kvpair.compress_threshold = acl->quotation.compress_threshold;

	//根据客户端ip网段匹配集群名以及节点IP
	for (j = 0;j < len;j++)
	{
		char p = data[j];
		rest[j] = p;
		if ( p == 46 ) num++;
		if ( num == 2 ) break;
	}
	key = std::string(rest,j);

	it_client = map_client.find(key);
	if ( it_client == map_client.end() ) {
		acl->route.is_match_ip = false;
		return QKPACK_OK;
	}

	it_server = map_server.find(it_client->second);
	if ( it_server != map_server.end() ) {
		acl->route.is_match_ip = true;
		return QKPACK_OK;
	}
	
	acl->route.is_match_ip = false;
	return QKPACK_OK;
}


qkpack_acl_t* QkPackACL::Process(qkpack_request_t *request) 
{
	int													rc;
	qkpack_acl_t										*acl = NULL;

	rc = SetNameSpace(request);
	if ( rc != QKPACK_OK ) {
		return NULL;
	}
	
	//set acl info
	acl = SetACL( request->ak, request->uri);
	if ( acl == NULL ) {
		request->desc = QKPACK_ERROR_ACL_NO_AUTH;
		return NULL;
	}
			
	rc = SetClusterName(request);
	if ( rc != QKPACK_OK ) {
		return NULL;
	}	
	
	//get node list
	rc = GetNodeList(request, acl);
	if ( rc != QKPACK_OK ) {
		return NULL;
	}
		
	//权限验证
	if ( !(acl->operation & request->kvpair.operation_type) ) {
	
		request->desc = (request->kvpair.operation_type == 
				QKPACK_OPERATION_READ?QKPACK_ERROR_INTERFACE_OPERATION_READ:QKPACK_ERROR_INTERFACE_OPERATION_WRITE);
		return NULL;
	}

	//check keylen,vallen,datalen
	rc = CheckInterfaceLimitsLength(request, acl->quotation.vector_limits_length);
	if ( rc != QKPACK_OK ) {
		return NULL;
	}

	//check zfixedsetadd limit
	rc = GetZsetLimit(request, acl->quotation.vector_policies);
	if ( rc != QKPACK_OK ) {
		return NULL;
	}
	
	//check interface timeout
	rc = CheckInterfaceTimeout(request, acl->quotation.vector_timeouts);
	if ( rc != QKPACK_OK ) {
		return NULL;
	}

	return acl;
}


qkpack_acl_t* QkPackACL::SetACL(const std::string &ak, const std::string &uri)
{
	int													rc;
	std::string											key = uri + "&&" + ak;
	std::map<std::string, qkpack_acl_t*>::const_iterator		it;

	it = map_acl_.find(key);		
	if( it != map_acl_.end() ) {
		return it->second;
	}
	
	qkpack_acl_t *acl = new qkpack_acl_t();
	rc = GetCurlResponse(acl,ak, uri); 
	if ( rc != QKPACK_OK ) {
		delete acl;
		return NULL;
	}
	map_acl_.insert(std::make_pair(key, acl));

	return acl;
}


int QkPackACL::GetCurlResponse(qkpack_acl_t *acl,const std::string &ak, const std::string &uri)
{
	int													rc;
	EasyCurl											easy_curl;
	std::vector<std::string>							headers;
	std::string											data;

	//POST /data-auth-http-server/httpServer/authority
	//Content-Type: application/json;charset=UTF-8
	//{"app_key":"GBeArhYgWl","uri":"qkd://BJ001/lj"}

	QKPACK_LOG_DEBUG("acl post /data-auth-http-server/httpServer/authority");
	
	headers.push_back("Content-Type: application/json;charset=UTF-8");
	data = "{\"app_key\":\""+ak+"\",\"uri\":\""+uri+"\"}";
	
	rc = easy_curl.Post(acl_url_, headers, data, timeout_);
	if ( rc != QKPACK_OK || !easy_curl.GetResponse().length() ) {
		return QKPACK_ERROR;
	}
	
	rc = qkpack_yajl_json_ ->ReadACLResponse(acl, easy_curl.GetResponse());
	if ( rc != QKPACK_OK ) {
		return QKPACK_ERROR;
	}

	return QKPACK_OK;
}

File: module_qkpack/qkpack/resty/qkpack_acl.lua
local bit = require "bit"
local ljson_decoder = require 'json_decoder'
local decoder = ljson_decoder.new()

--local cjson_safe = require "cjson.safe"
local qkpack_common = require "resty.qkpack_common"

local band = bit.band
local len = string.len
local find = string.find
local tonumber = tonumber

local new_tab = require "table.new"

local _M = {
    _VERSION = '0.01',
}

_M.acl_url = "http://10.100.14.152:8080/data-auth-http-server/httpServer/authority"
_M.table_acl = {}


local function split(str, sep, plain)

	local								b, res = 0, {}
	sep = sep or '%s+'

	assert(type(sep) == 'string')
	assert(type(str) == 'string')

	if #sep == 0 then
		for i = 1, #str do
			res[#res + 1] = string.sub(str, i, i)
		end
		return res
	end

	while b <= #str do
		local e, e2 = string.find(str, sep, b, plain)
		if e then
			res[#res + 1] = string.sub(str, b, e-1)
			b = e2 + 1
			if b > #str then res[#res + 1] = "" end
		else
			res[#res + 1] = string.sub(str, b)
			break
		end
	end
	return res
end


local function set_namespace(self, request, uri)

	local								rc = 0
	local								t = nil
	local								f = nil
	local								namespace = nil
	
	f,_ = find(tostring(uri), tostring(qkpack_common.QKPACK_ACL_URI_FORMAT))
	if f == nil then
		request.desc = qkpack_common.QKPACK_ERROR_ACL_FORMAT
		return qkpack_common.QKPACK_ERROR
	end

	t = split(uri, "/", true)
	if type(t) ~= "table" or table.getn(t) ~= 4 then
		request.desc = qkpack_common.QKPACK_ERROR_ACL_NAMESPACE_NOT_EMPTY
		return qkpack_common.QKPACK_ERROR
	end

	namespace = t[4]

	if  len(namespace) ~= qkpack_common.QKPACK_ACL_NAMESPACE_LENGTH  then
		request.desc = qkpack_common.QKPACK_ERROR_ACL_NAMESPACE_LENGTH;
		return qkpack_common.QKPACK_ERROR;
	end

	request.cluster_name = t[3]
	request.namespace = namespace

	return qkpack_common.QKPACK_OK	
end


local function get_node_list(self, request, route, acl)
	
	local								rc = 0
	local								t = nil
	local								x_real_ip =  ngx.var.server_addr
	local								key = nil
	local								node_list = nil

	if acl.node_list ~= nil then
		return qkpack_common.QKPACK_OK
	end

	acl.node_list = {}


	if route == nil or route == ngx.null or type(route) ~= "table" then
		ngx.log(ngx.ERR, "acl route is nil")
		return qkpack_common.QKPACK_ERROR
	end

	t = split(x_real_ip, ".", true)
	if t == nil then
		ngx.log(ngx.ERR, "split t is nil")
		return qkpack_common.QKPACK_ERROR
	end
	
	key =  t[1].."."..t[2]

	local temp_name = nil
	for k,v in pairs(route) do
		if k == "client" then
			for k1,v1 in pairs(v) do
				if k1 == key then
					temp_name = v1
					break
				end
			end
		end
	end
	
	if temp_name == nil then
	
		for k,v in pairs(route) do
			if k == "default" then
				node_list = v
				break
			end
		end
	
	else
		for k,v in pairs(route) do
			if k == "server" then
				for k1,v1 in pairs(v) do
					if k1 == temp_name then
						node_list = v1
						break
					end
				end
			end
		end
	end
	
	if node_list == nil then
		return qkpack_common.QKPACK_ERROR
	end

	t = split(node_list, ",", true)
	for _,v in pairs(t) do
		local node = split(v, ":", true)
		table.insert(acl.node_list, { node[1], tonumber(node[2]) } )	
	end

	request.node_list = acl.node_list
	return qkpack_common.QKPACK_OK
end

--get exptime
local function get_exptime(self, request, quotation)

	if quotation == nil then
		return qkpack_common.QKPACK_ERROR
	end

	for k,v in pairs(quotation) do

		if k == "expire" then
			request.exptime = v
		end

		if k == "compressThreshold" then
			request.compress_threshold = v
		end
	end
	
	if request.exptime == nil or request.exptime <= 0 then
		return qkpack_common.QKPACK_ERROR
	end
	
	
	
	return qkpack_common.QKPACK_OK
end

--set acl
local function set_acl(self, uri, ak )
		
	local								rc = 0
	local								key = nil
	local								data

	if uri == nil or ak == nil then
		return nil
	end


	key = uri.."&&"..ak
	if self.table_acl[key] ~= nil then
		return self.table_acl[key]
	end

	data = "{\"app_key\":\""..ak.."\",\"uri\":\""..uri.."\"}"
	
	local http = require "resty.http"
	local httpc = http.new()

	local res, err = httpc:request_uri(self.acl_url, {
		method = "POST",
		body = data,
		headers = {
			["Content-Type"] = "application/json;charset=UTF-8",
		}
	})

	if res == nil or res.status ~= 200 or res.body == nil  then
		return nil
	end
	

	local buffer, err = decoder:decode(res.body)
	if buffer == nil then
		return nil
	end
	
	if buffer.uri_id == -1 then
		return nil
	end

	self.table_acl[key] = buffer
	return self.table_acl[key]
end



--get_zset_limit
function _M.get_zset_limit(self, uri, ak, ken_len)
	
	--ngx.log(ngx.DEBUG, "get_zset_limit begin---------------------------------------------------")

	local								acl = nil
	local								quotation = nil
	local								status = false	
	local								limit = 0
	local								k_len = tonumber(ken_len)

	--request acl
	acl = set_acl(self, uri, ak)
	if acl == nil then
		return qkpack_common.QKPACK_ERROR, "acl is nil"
	end

	quotation = acl.quotation	
	if quotation == nil then
		return qkpack_common.QKPACK_ERROR, "quotation is nil"
	end

	for k,v in pairs(quotation) do
		if k == "limitPolicies" then

			for _,v1 in pairs(v) do
				
				if k_len >= v1["min"] and k_len <= v1["max"] then	
					status = true
					limit = v1["limit"]
					break
				end
			
			end
		
		end
	end

	if status == false then
		return qkpack_common.QKPACK_ERROR, qkpack_common.QKPACK_ERROR_ACL_LIMITPOLICIES_LIMIT
	end
	
	--ngx.log(ngx.DEBUG, "get_zset_limit end---------------------------------------------------")

	return qkpack_common.QKPACK_OK, limit
end

function _M.set_uri_id(self, request)
	
    	local                               				rc = 0	
	local								acl = nil
	local								data = nil

	if request.kvpair ~= nil then
		data = request.kvpair
	elseif request.multikvpair ~= nil then
		data = request.multikvpair
	elseif request.sset_member ~= nil then
		data = request.sset_member
	elseif request.zset_member ~= nil then
		data = request.zset_member
	elseif request.zset_query ~= nil then
		data = request.zset_query
	end

	if data == nil then
		ngx.log(ngx.ERR, "data is nil")
		return qkpack_common.QKPACK_ERROR
	end
	
	--set namespace,cluster_name
	rc =  set_namespace(self, request, data.uri)
	if rc ~= qkpack_common.QKPACK_OK then
		ngx.log(ngx.ERR, "set_namespace is error")
		return qkpack_common.QKPACK_ERROR
	end

	--request acl
	acl = set_acl(self, data.uri, data.ak)
	if acl == nil then
		ngx.log(ngx.ERR, "set_acl is nil")
		request.desc = qkpack_common.QKPACK_ERROR_ACL_NO_AUTH
		return qkpack_common.QKPACK_ERROR
	end

	request.uri_id = acl.uri_id
	return qkpack_common.QKPACK_OK
end




function _M.process(self, request)
	
	--ngx.log(ngx.DEBUG, "process begin---------------------------------------------------")

    	local                               				rc = 0	
	local								acl = nil
	local								data = nil

	if request.kvpair ~= nil then
		data = request.kvpair
	elseif request.multikvpair ~= nil then
		data = request.multikvpair
	elseif request.sset_member ~= nil then
		data = request.sset_member
	elseif request.zset_member ~= nil then
		data = request.zset_member
	elseif request.zset_query ~= nil then
		data = request.zset_query
	end

	if data == nil then
		ngx.log(ngx.ERR, "data is nil")
		return qkpack_common.QKPACK_ERROR
	end

	--set namespace,cluster_name
	rc =  set_namespace(self, request, data.uri)
	if rc ~= qkpack_common.QKPACK_OK then
		ngx.log(ngx.ERR, "set_namespace is error")
		return qkpack_common.QKPACK_ERROR
	end

	--request acl
	acl = set_acl(self, data.uri, data.ak)
	if acl == nil then
		ngx.log(ngx.ERR, "set_acl is nil")
		request.desc = qkpack_common.QKPACK_ERROR_ACL_NO_AUTH
		return qkpack_common.QKPACK_ERROR
	end

	--get node list
	rc = get_node_list(self, request, acl.route, acl)
	if rc ~= qkpack_common.QKPACK_OK then
		ngx.log(ngx.ERR, "get_node_list is error")
		request.desc = qkpack_common.QKPACK_ERROR_ACL_NO_AUTH
		return qkpack_common.QKPACK_ERROR
	end


	--auth
	request.uri_id = acl.uri_id
	local auth = band(acl.operation , request.operation_type) 
	if  (auth == 0) or (not auth) then
	
		request.desc = 	
		request.operation_type == qkpack_common.QKPACK_OPERATION_READ and 
		qkpack_common.QKPACK_ERROR_INTERFACE_OPERATION_READ or 
		qkpack_common.QKPACK_ERROR_INTERFACE_OPERATION_WRITE

		return qkpack_common.QKPACK_ERROR
	end

	--get exptime
	rc = get_exptime(self, request, acl.quotation)
	if rc ~= qkpack_common.QKPACK_OK then
		ngx.log(ngx.ERR, "get_node_list is error")
		return qkpack_common.QKPACK_ERROR
	end

	--todo check interface timeout

	--ngx.log(ngx.DEBUG, "process end---------------------------------------------------")
	
	return qkpack_common.QKPACK_OK	
end


return _M
